common/kotlinx-coroutines-core-common/src/main/kotlin/kotlinx/coroutines/experimental/channels/AbstractChannel.kt

```diff
@@ -243,23 +243,53 @@ public abstract class AbstractSendChannel<E> : SendChannel<E> {
 
     public override fun close(cause: Throwable?): Boolean {
         val closed = Closed<E>(cause)
+
+        /*
+         * Try to commit close by adding a close token to the end of the queue.
+         * Successful -> we're now responsible for closing receivers
+         * Not successful -> help closing pending receivers to maintain invariant
+         * "if (!close()) next send will throw"
+         */
+        val closeAdded = queue.addLastIfPrev(closed, { it !is Closed<*> })
+        if (!closeAdded) {
+            helpClose(queue.prevNode as Closed<*>)
+            return false
+        }
+
+        helpClose(closed)
+        onClosed(closed)
+        afterClose(cause)
+        // Cancel it as the last action so if the channel is closed, then the job is cancelled as well
+        job.cancel(cause)
+        return true
+    }
+
+    private fun helpClose(closed: Closed<*>) {
```

Comment:
Discuss: `helpClose` should probably try to cancel job as well, so it will be linearizable along with channel

Reply:
Given the current implementation doesn't try to cancel the job, could this lead to a deadlock if somehow, a `runBlocking { ... }` call is waiting for it because it'd be waiting for the job to complete?

Reply:
The current implementation doesn't have a job, this branch is not going to `master`.